<!DOCTYPE html>
<html>
  <head>
    <title>Beyond exception handling &middot; Advanced R.</title>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <link href="www/bootstrap.min.css" rel="stylesheet">
    <link href="www/highlight.css" rel="stylesheet">

    <link href='http://fonts.googleapis.com/css?family=Inconsolata:400,700'
      rel='stylesheet' type='text/css'>
  </head>

  <body>

    <div class="container">

      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li class="dropdown">
            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
              Table of contents<b class="caret"></b>
            </a>
            <ul class="dropdown-menu pull-right" role="menu">
              <li><a href="Introduction.html">Introduction</a></li>
<li class="dropdown-header">Foundations</li>
<li><a href="Data-structures.html">Data structures</a></li>
<li><a href="Subsetting.html">Subsetting</a></li>
<li><a href="Vocabulary.html">Vocabulary</a></li>
<li><a href="Style.html">Style</a></li>
<li><a href="Functions.html">Functions</a></li>
<li><a href="OO-essentials.html">OO field guide</a></li>
<li><a href="Environments.html">Environments</a></li>
<li><a href="Exceptions-Debugging.html">Exceptions and debugging</a></li>

<li class="dropdown-header">Functional programming</li>
<li><a href="Functional-programming.html">Functional programming</a></li>
<li><a href="Functionals.html">Functionals</a></li>
<li><a href="Function-operators.html">Function operators</a></li>

<li class="dropdown-header">Metaprogramming</li>
<li><a href="Computing-on-the-language.html">Non-standard evaluation</a></li>
<li><a href="Expressions.html">Expressions</a></li>
<li><a href="dsl.html">Domain specific languages</a></li>

<li class="dropdown-header">Performant code</li>
<li><a href="Performance.html">Performance</a></li>
<li><a href="Profiling.html">Profiling</a></li>
<li><a href="memory.html">Memory</a></li>
<li><a href="Rcpp.html">Rcpp</a></li>
<li><a href="C-interface.html">R's C interface</a></li>

            </ul>
          </li>
        </ul>

        <h3 class="muted"><a href="/">Advanced R</a> <small>by Hadley Wickham</small></h3>
        <hr>
      </div>

      <div class="row">
        <div class="col-xs-12 col-sm-3" id="nav">
        <div class="well">
          Want to learn from me in person? I'm next teaching in <a href="https://rstudio-chicago.eventbrite.com">Chicago, May 27-28</a>.
        </div>

        <div class="well">
          Want a physical copy of this material? <a href="http://amzn.com/1466586966?tag=devtools-20">Buy a book from amazon!</a>.
        </div>

        <h4>Contents</h4>
          <ul class="list-unstyled" id="toc"></ul>

          <hr>
          <p><a href="/contribute.html">How to contribute</a></p>

          <p><a class="btn btn-primary" href="https://github.com/hadley/adv-r/edit/master/beyond-exception-handling.rmd">Edit this page</a></p>
        </div>

        <div id="content" class="col-xs-12 col-sm-8 pull-right">
          <p>This is a translation of <a href="http://www.gigamonkeys.com/book/beyond-exception-handling-conditions-and-restarts.html">Beyond exception handling: conditions and restarts</a> by Peter Seibel, from Lisp to R. The original document is copyright (c) 2003-2005, Peter Seibel; translated with permission.</p>
<p>The majority of the translation involves changing Lisp syntax to R syntax. There are few differences in the overall system.</p>
<h1>Beyond Exception Handling: Conditions and Restarts</h1>
<p>One of R’s great features is its <em>condition</em> system. It serves a similar purpose to the exception handling systems in Java, Python, and C++ but is more flexible. In fact, its flexibility extends beyond error handling–conditions are more general than exceptions in that a condition can represent any occurrence during a program’s execution that may be of interest to code at different levels on the call stack. For example, in the section “Other Uses for Conditions,” you’ll see that conditions can be used to emit warnings without disrupting execution of the code that emits the warning while allowing code higher on the call stack to control whether the warning message is printed. For the time being, however, I’ll focus on error handling.</p>
<p>The condition system is more flexible than exception systems because instead of providing a two-part division between the code that signals an error<sup>1</sup> and the code that handles it,<sup>2</sup> the condition system splits the responsibilities into three parts–<em>signaling</em> a condition, <em>handling</em> it, and <em>restarting</em>. In this chapter, I’ll describe how you could use conditions in part of a hypothetical application for analyzing log files. You’ll see how you could use the condition system to allow a low-level function to detect a problem while parsing a log file and signal an error, to allow mid-level code to provide several possible ways of recovering from such an error, and to allow code at the highest level of the application to define a policy for choosing which recovery strategy to use.</p>
<p>To start, I’ll introduce some terminology: <em>errors</em>, as I’ll use the term, are the consequences of Murphy’s law. If something can go wrong, it will: a file that your program needs to read will be missing, a disk that you need to write to will be full, the server you’re talking to will crash, or the network will go down. If any of these things happen, it may stop a piece of code from doing what you want. But there’s no bug; there’s no place in the code that you can fix to make the nonexistent file exist or the disk not be full. However, if the rest of the program is depending on the actions that were going to be taken, then you’d better deal with the error somehow or you <em>will</em> have introduced a bug. So, errors aren’t caused by bugs, but neglecting to handle an error is almost certainly a bug.</p>
<p>So, what does it mean to handle an error? In a well-written program, each function is a black box hiding its inner workings. Programs are then built out of layers of functions: high-level functions are built on top of the lower-level functions, and so on. This hierarchy of functionality manifests itself at runtime in the form of the call stack: if <code>high</code> calls <code>medium</code>, which calls <code>low</code>, when the flow of control is in <code>low</code>, it’s also still in <code>medium</code> and <code>high</code>, that is, they’re still on the call stack.</p>
<p>Because each function is a black box, function boundaries are an excellent place to deal with errors. Each function–<code>low</code>, for example–has a job to do. Its direct caller–<code>medium</code> in this case–is counting on it to do its job. However, an error that prevents it from doing its job puts all its callers at risk: <code>medium</code> called <code>low</code> because it needs the work done that <code>low</code> does; if that work doesn’t get done, <code>medium</code> is in trouble. But this means that <code>medium</code>’s caller, <code>high</code>, is also in trouble–and so on up the call stack to the very top of the program. On the other hand, because each function is a black box, if any of the functions in the call stack can somehow do their job despite underlying errors, then none of the functions above it needs to know there was a problem–all those functions care about is that the function they called somehow did the work expected of it.</p>
<p>In most languages, errors are handled by returning from a failing function and giving the caller the choice of either recovering or failing itself. Some languages use the normal function return mechanism, while languages with exceptions return control by <em>throwing</em> or <em>raising</em> an exception. Exceptions are a vast improvement over using normal function returns, but both schemes suffer from a common flaw: while searching for a function that can recover, the stack unwinds, which means code that might recover has to do so without the context of what the lower-level code was trying to do when the error actually occurred.</p>
<p>Consider the hypothetical call chain of <code>high</code>, <code>medium</code>, <code>low</code>. If <code>low</code> fails and <code>medium</code> can’t recover, the ball is in <code>high</code>’s court. For <code>high</code> to handle the error, it must either do its job without any help from <code>medium</code> or somehow change things so calling <code>medium</code> will work and call it again. The first option is theoretically clean but implies a lot of extra code–a whole extra implementation of whatever it was <code>medium</code> was supposed to do. And the further the stack unwinds, the more work that needs to be redone. The second option–patching things up and retrying–is tricky; for <code>high</code> to be able to change the state of the world so a second call into <code>medium</code> won’t end up causing an error in <code>low</code>, it’d need an unseemly knowledge of the inner workings of both <code>medium</code> and <code>low</code>, contrary to the notion that each function is a black box.</p>
<h2>The R Way</h2>
<p>R’s error handling system gives you a way out of this conundrum by letting you separate the code that actually recovers from an error from the code that decides how to recover. Thus, you can put recovery code in low-level functions without committing to actually using any particular recovery strategy, leaving that decision to code in high-level functions.</p>
<p>To get a sense of how this works, let’s suppose you’re writing an application that reads some sort of textual log file, such as a Web server’s log. Somewhere in your application you’ll have a function to parse the individual log entries. Let’s assume you’ll write a function, <code>parse_log_entry()</code>, that will be passed a string containing the text of a single log entry and that is supposed to return a <code>log_entry()</code> object representing the entry. This function will be called from a function, <code>parse_log_file()</code>, that reads a complete log file and returns a list of objects representing all the entries in the file.</p>
<p>To keep things simple, the <code>parse_log_entry()</code> function will not be required to parse incorrectly formatted entries. It will, however, be able to detect when its input is malformed. But what should it do when it detects bad input? In C you’d return a special value to indicate there was a problem. In Java or Python you’d throw or raise an exception. In R, you signal a condition.</p>
<h2>Conditions</h2>
<p>A <em>condition</em> is an S3 object whose class indicates the general nature of the condition and whose instance data carries information about the details of the particular circumstances that lead to the condition being signaled.<sup>3</sup> In this hypothetical log analysis program, you might define a condition class, <code>malformed_log_entry_error</code>, that <code>parse_log_entry()</code> will signal if it’s given data it can’t parse.</p>
<p>Conditional classes are regular S3 classes, built up from a list with components <code>message</code> and <code>call</code>. There is no built in function to generate a new object of class condition, but we can add one:</p>
<pre class="sourceCode r"><code class="sourceCode r">condition &lt;-<span class="st"> </span>function(subclass, message, <span class="dt">call =</span> <span class="kw">sys.call</span>(-<span class="dv">1</span>), ...) {
  <span class="kw">structure</span>(
    <span class="dt">class =</span> <span class="kw">c</span>(subclass, <span class="st">&quot;condition&quot;</span>),
    <span class="kw">list</span>(<span class="dt">message =</span> message, <span class="dt">call =</span> call, ...)
  )
}</code></pre>
<p>When using the condition system for error handling, you should define your conditions as subclasses of <code>error</code>, a subclass of <code>condition</code>. Thus, you might define <code>malformed_log_entry_error</code>, with a slot to hold the argument that was passed to <code>parse_log_entry()</code>, like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">malformed_log_entry_error &lt;-<span class="st"> </span>function(text) {
  msg &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Malformed log entry: &quot;</span>, text)
  <span class="kw">condition</span>(<span class="kw">c</span>(<span class="st">&quot;malformed_log_entry_entry&quot;</span>, <span class="st">&quot;error&quot;</span>),
    <span class="dt">message =</span> msg, 
    <span class="dt">text =</span> text
  )
}</code></pre>
<h2>Condition Handlers</h2>
<p>In <code>parse_log_entry()</code> you’ll signal a <code>malformed_log_entry_error</code> if you can’t parse the log entry. You signal errors with the function <code>stop()</code>. <code>stop()</code> is normally just called with a string, the error message, but you can also call it with a condition object. Thus, you could write <code>parse_log_entry()</code> like this, eliding the details of actually parsing a log entry:</p>
<pre class="sourceCode r"><code class="sourceCode r">parse_log_entry &lt;-<span class="st"> </span>function(text) {
  if (!<span class="kw">well_formed_log_entry</span>(text)) {
    <span class="kw">stop</span>(<span class="kw">malformed_log_entry_error</span>(text))
  }
  
  <span class="kw">new_log_entry</span>(text)
}</code></pre>
<p>What happens when the error is signaled depends on the code above <code>parse_log_entry()</code> on the call stack. To avoid a top level error message, you must establish a <em>condition handler</em> in one of the functions leading to the call to <code>parse_log_entry</code>. When a condition is signaled, the signaling machinery looks through a list of active condition handlers, looking for a handler that can handle the condition being signaled based on the condition’s class. Each condition handler consists of a type specifier indicating what types of conditions it can handle and a function that takes a single argument, the condition. At any given moment there can be many active condition handlers established at various levels of the call stack. When a condition is signaled, the signaling machinery finds the most recently established handler whose type specifier is compatible with the condition being signaled and calls its function, passing it the condition object.</p>
<p>The handler function can then choose whether to handle the condition. The function can decline to handle the condition by simply returning normally, in which case control returns to next most recently established handler with a compatible type specifier. To handle the condition, the function must transfer control out of the signaller via a <em>nonlocal exit</em>. In the next section, you’ll see how a handler can choose where to transfer control. However, many condition handlers simply want to unwind the stack to the place where they were established and then run some code. The function <code>tryCatch()</code> establishes this kind of condition handler. The basic form of a <code>tryCatch</code> is as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tryCatch</span>(expression, 
  <span class="dt">condition_class_1 =</span> function(var) ...,
  <span class="dt">condition_class_2 =</span> function(var) ...,
  ...
)</code></pre>
<p>If the <em>expression</em> returns normally, then its value is returned by the <code>tryCatch()</code>. The body of a <code>tryCatch()</code> must be a single expression; but you can always use <code>{</code> to combine several expressions into a single form. If, however, the expression signals a condition that’s an instance of any of the <em>condition class</em>s then the code in the appropriate error clause is executed and its value returned by the <code>tryCatch()</code>. The <em>var</em> is the name of the variable that will hold the condition object when the handler code is executed. If the code doesn’t need to access the condition object, you can omit the variable name.</p>
<p>For instance, one way to handle the <code>malformed_log_entry_error</code> signaled by <code>parse_log_entry()</code> in its caller, <code>parse_log_file()</code>, would be to skip the malformed entry. In the following function, the <code>tryCatch()</code> expression will either return the value returned by <code>parse_log_entry()</code> or return <code>NULL</code> if a <code>malformed_log_entry_error</code> is signaled.</p>
<pre class="sourceCode r"><code class="sourceCode r">parse_log_file &lt;-<span class="st"> </span>function(file) {
  lines &lt;-<span class="st"> </span><span class="kw">readLines</span>(file)
  
  <span class="kw">lapply</span>(lines, function(text) {
    <span class="kw">tryCatch</span>(
      <span class="dt">malformed_log_entry =</span> function(e) <span class="ot">NULL</span>,
      <span class="kw">parse_log_entry</span>(text)
    )
  })
}</code></pre>
<p>When <code>parse_log_entry()</code> returns normally, its value will be collected by the <code>lapply()</code>. But if <code>parse_log_entry</code> signals a <code>malformed_log_entry_error</code>, then the error clause will return <code>NULL</code>.</p>
<p>This version of <code>parse_log_file()</code> has one serious deficiency: it’s doing too much. As its name suggests, the job of <code>parse_log_file()</code> is to parse the file and produce a list of <code>log_entry</code> objects; if it can’t, it’s not its place to decide what to do instead. What if you want to use <code>parse_log_file()</code> in an application that wants to tell the user that the log file is corrupted or one that wants to recover from malformed entries by fixing them up and re-parsing them? Or maybe an application is fine with skipping them but only until a certain number of corrupted entries have been seen.</p>
<p>You could try to fix this problem by moving the <code>tryCatch()</code> to a higher-level function. However, then you’d have no way to implement the current policy of skipping individual entries–when the error was signaled, the stack would be unwound all the way to the higher-level function, abandoning the parsing of the log file altogether. What you want is a way to provide the current recovery strategy without requiring that it always be used.</p>
<h3>Java style exception handling</h3>
<p><code>tryCatch</code> is the nearest analog in R to Java- or Python-style exception handling. Where you might write this in Java:</p>
<pre><code>try {
  doStuff();
  doMoreStuff();
} catch (SomeException se) {
  recover(se);
}</code></pre>
<p>or this in Python:</p>
<pre><code>try:
  doStuff()
  doMoreStuff()
except SomeException, se:
  recover(se)</code></pre>
<p>in R you’d write this:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tryCatch</span>({
  <span class="kw">doStuff</span>()
  <span class="kw">doMoreStuff</span>()
}, <span class="dt">some_exception =</span> function(se) {
  <span class="kw">recover</span>(se)
})</code></pre>
<h2>Restarts</h2>
<p>The condition system lets you do this by splitting the error handling code into two parts. You place code that actually recovers from errors into <em>restarts</em>, and condition handlers can then handle a condition by invoking an appropriate restart. You can place restart code in mid- or low-level functions, such as <code>parse_log_file()</code> or <code>parse_log_entry()</code>, while moving the condition handlers into the upper levels of the application.</p>
<p>To change <code>parse_log_file()</code> so it establishes a restart instead of a condition handler, you can change the <code>tryCatch()</code> to a <code>withRestarts()</code>. The form of <code>withRestarts</code> is very similar to a <code>tryCatch()</code>. In general, a restart name should describe the action the restart takes. In <code>parse_log_file()</code>, you can call the restart <code>skip_log_entry</code> since that’s what it does. The new version will look like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">parse_log_file &lt;-<span class="st"> </span>function(file) {
  lines &lt;-<span class="st"> </span><span class="kw">readLines</span>(file)
  
  <span class="kw">lapply</span>(lines, function(text) {
    <span class="kw">withRestarts</span>(
      <span class="kw">parse_log_entry</span>(text),
      <span class="dt">skip_log_entry =</span> function(e) <span class="ot">NULL</span>
    )
  })
}</code></pre>
<p>If you invoke this version of <code>parse_log_file()</code> on a log file containing corrupted entries, it won’t handle the error directly; you’ll end up in the debugger. However, there among the various restarts listed by <code>findRestarts()</code> will be one called <code>skip_log_entry</code>, which, if you choose it, will cause <code>parse_log_file()</code> to continue on its way as before. To avoid ending up in the debugger, you can establish a condition handler that invokes the <code>skip_log_entry</code> restart automatically.</p>
<p>The advantage of establishing a restart rather than having <code>parse_log_file()</code> handle the error directly is it makes <code>parse_log_file()</code> usable in more situations. The higher-level code that invokes <code>parse_log_file()</code> doesn’t have to invoke the <code>skip_log_entry</code> restart. It can choose to handle the error at a higher level. Or, as I’ll show in the next section, you can add restarts to <code>parse_log_entry()</code> to provide other recovery strategies, and then condition handlers can choose which strategy they want to use.</p>
<p>But before I can talk about that, you need to see how to set up a condition handler that will invoke the <code>skip_log_entry</code> restart. You can set up the handler anywhere in the chain of calls leading to <code>parse_log_file()</code>. This may be quite high up in your application, not necessarily in <code>parse_log_file()</code>’s direct caller. For instance, suppose the main entry point to your application is a function, <code>log_analyzer()</code>, that finds a bunch of logs and analyzes them with the function <code>analyze_log()</code>, which eventually leads to a call to <code>parse_log_file()</code>. Without any error handling, it might look like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">log_analyzer &lt;-<span class="st"> </span>function() {
  logs &lt;-<span class="st"> </span><span class="kw">find_all_logs</span>()
  <span class="kw">lapply</span>(logs, analyze_log)
}</code></pre>
<p>The job of <code>analyze_log()</code> is to call, directly or indirectly, <code>parse_log_file()</code> and then do something with the list of log entries returned. An extremely simple version might look like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">analyze_log &lt;-<span class="st"> </span>function(log) {
  entries &lt;-<span class="st"> </span><span class="kw">parse_log_file</span>(log)
  <span class="kw">lapply</span>(entries, anaylze_entry)
}</code></pre>
<p>where the function <code>analyze_entry()</code> is presumably responsible for extracting whatever information you care about from each log entry and stashing it away somewhere.</p>
<p>Assuming you always want to skip malformed log entries, you could change this function to establish a condition handler that invokes the <code>skip_log_entry</code> restart for you. However, you can’t use <code>tryCatch()</code> to establish the condition handler because then the stack would be unwound to the function where the <code>tryCatch()</code> appears. Instead, you need to use the lower-level macro <code>withCallingHandlers()</code>. The basic form of <code>withCallingHandlers()</code> is as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">withCallingHandlers</span>(
  expr,
  <span class="dt">condition_1 =</span> function() ...,
  <span class="dt">condition_2 =</span> function() ....
)</code></pre>
<p>An important difference between <code>tryCatch()</code> and <code>withCallingHandlers()</code> is that the handler function bound by <code>withCallingHandlers()</code> will be run without unwinding the stack–the flow of control will still be in the call to <code>parse_log_entry()</code> when this function is called. A call to <code>invokeRestart()</code> will find and invoke the most recently bound restart with the given name. So you can add a handler to <code>log_analyzer()</code> that will invoke the <code>skip_log_entry</code> restart established in <code>parse_log_file()</code> like this:<sup>5</sup></p>
<pre class="sourceCode r"><code class="sourceCode r">log_analyzer &lt;-<span class="st"> </span>function() {
  logs &lt;-<span class="st"> </span><span class="kw">find_all_logs</span>()
  
  <span class="kw">withCallingHandlers</span>(
    <span class="dt">malformed_log_entry_error =</span> function(e) <span class="kw">invokeRestart</span>(<span class="st">&quot;skip_log_entry&quot;</span>),
    <span class="kw">lapply</span>(logs, analyze_log)
  )
}</code></pre>
<p>In this <code>withCallingHandlers()</code>, the handler function is an anonymous function that invokes the restart <code>skip_log_entry</code>. You could also define a named function that does the same thing and bind it instead. In fact, a common practice when defining a restart is to define a function, with the same name and taking a single argument, the condition, that invokes the eponymous restart. Such functions are called <em>restart functions</em>. You could define a restart function for <code>skip_log_entry</code> like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">skip_log_entry &lt;-<span class="st"> </span>function() <span class="kw">invokeRestart</span>(<span class="st">&quot;skip_log_entry&quot;</span>)</code></pre>
<p>Then you could change the definition of <code>log_analyzer()</code> to this:</p>
<pre class="sourceCode r"><code class="sourceCode r">log_analyzer &lt;-<span class="st"> </span>function() {
  logs &lt;-<span class="st"> </span><span class="kw">find_all_logs</span>()
  
  <span class="kw">withCallingHandlers</span>(
    <span class="dt">malformed_log_entry_error =</span> skip_log_entry,
    <span class="kw">lapply</span>(logs, analyze_log)
  )
}</code></pre>
<p>As written, the <code>skip_log_entry</code> restart function assumes that a <code>skip_log_entry</code> restart has been established. If a <code>malformed_log_entry_error</code> is ever signaled by code called from <code>log_analyzer()</code> without a <code>skip_log_entry</code> having been established, the call to <code>invokeRestart()</code> will signal an error when it fails to find the <code>skip_log_entry</code> restart. If you want to allow for the possibility that a <code>malformed_log_entry_error</code> might be signaled from code that doesn’t have a <code>skip_log_entry</code> restart established, you could change the <code>skip_log_entry</code> function to this:</p>
<pre class="sourceCode r"><code class="sourceCode r">skip_log_entry &lt;-<span class="st"> </span>function() {
  r &lt;-<span class="st"> </span><span class="kw">findRestart</span>(<span class="st">&quot;skip_log_entry&quot;</span>) 
  if (<span class="kw">is.null</span>(r)) <span class="kw">return</span>()
  
  <span class="kw">invokeRestart</span>(r)
}</code></pre>
<p><code>findRestart</code> looks for a restart with a given name and returns an object representing the restart if the restart is found and <code>NULL</code> if not. You can invoke the restart by passing the restart object to <code>invokeRestart()</code>. Thus, when <code>skip_log_entry</code> is bound with <code>withCallingHandlers()</code>, it will handle the condition by invoking the <code>skip_log_entry</code> restart if one is available and otherwise will return normally, giving other condition handlers, bound higher on the stack, a chance to handle the condition.</p>
<h2>Providing Multiple Restarts</h2>
<p>Since restarts must be explicitly invoked to have any effect, you can define multiple restarts, each providing a different recovery strategy. As I mentioned earlier, not all log-parsing applications will necessarily want to skip malformed entries. Some applications might want <code>parse_log_file()</code> to include a special kind of object representing malformed entries in the list of <code>log-entry</code> objects; other applications may have some way to repair a malformed entry and may want a way to pass the fixed entry back to <code>parse_log_entry()</code>.</p>
<p>To allow more complex recovery protocols, restarts can take arbitrary arguments, which are passed in the call to <code>invokeRestart()</code>. You can provide support for both the recovery strategies I just mentioned by adding two restarts to <code>parse_log_entry()</code>, each of which takes a single argument. One simply returns the value it’s passed as the return value of <code>parse_log_entry()</code>, while the other tries to parse its argument in the place of the original log entry.</p>
<pre class="sourceCode r"><code class="sourceCode r">parse_log_entry &lt;-<span class="st"> </span>function(text) {
  if (<span class="kw">well_formed_log_entry</span>(text)) {
    <span class="kw">return</span>(<span class="kw">new_log_entry</span>(text))
  }
  
  <span class="kw">withRestarts</span>(
    <span class="kw">stop</span>(<span class="kw">malformed_log_entry_error</span>(text)),
    <span class="dt">use_value =</span> function(x) x,
    <span class="dt">reparse_entry =</span> function(fixed_text) <span class="kw">parse_log_entry</span>(fixed_text)
  )
}</code></pre>
<p>The name <code>use_value()</code> is a standard name for this kind of restart. You can define a restart function for <code>use_value</code> similar to the <code>skip_log_entry</code> function you just defined.</p>
<pre class="sourceCode r"><code class="sourceCode r">use_value &lt;-<span class="st"> </span>function(x) <span class="kw">invokeRestart</span>(<span class="st">&quot;use_value&quot;</span>, x)</code></pre>
<p>So, if you wanted to change the policy on malformed entries to one that created an instance of <code>malformed_log_entry</code>, you could change <code>log_analyzer()</code> to this (assuming the existence of a <code>malformed_log_entry</code> constructor with a <code>text</code> parameter):</p>
<pre class="sourceCode r"><code class="sourceCode r">log_analyzer &lt;-<span class="st"> </span>function() {
  logs &lt;-<span class="st"> </span><span class="kw">find_all_logs</span>()
  
  <span class="kw">withCallingHandlers</span>(
    <span class="dt">malformed_log_entry_error =</span> function(text) {
      <span class="kw">use_value</span>(<span class="kw">malformed_log_entry</span>(text))
    },
    <span class="kw">lapply</span>(logs, analyze_log)
  )
}</code></pre>
<p>You could also have put these new restarts into <code>parse_log_file()</code> instead of <code>parse_log_entry()</code>. However, you generally want to put restarts in the lowest-level code possible. It wouldn’t, though, be appropriate to move the <code>skip_log_entry</code> restart into <code>parse_log_entry()</code> since that would cause <code>parse_log_entry()</code> to sometimes return normally with <code>NULL</code>, the very thing you started out trying to avoid. And it’d be an equally bad idea to remove the <code>skip_log_entry</code> restart on the theory that the condition handler could get the same effect by invoking the <code>use-value</code> restart with <code>NULL</code> as the argument; that would require the condition handler to have intimate knowledge of how the <code>parse_log_file()</code> works. As it stands, the <code>skip_log_entry</code> is a properly abstracted part of the log-parsing API.</p>
<h2>Other Uses for Conditions</h2>
<p>While conditions are mainly used for error handling, they can be used for other purposes–you can use conditions, condition handlers, and restarts to build a variety of protocols between low- and high-level code. The key to understanding the potential of conditions is to understand that merely signaling a condition has no effect on the flow of control.</p>
<p>The primitive signaling function <code>signalCondition()</code> implements the mechanism of searching for an applicable condition handler and invoking its handler function. The reason a handler can decline to handle a condition by returning normally is because the call to the handler function is just a regular function call–when the handler returns, control passes back to <code>signalCondition()</code>, which then looks for another, less recently bound handler that can handle the condition. If <code>signalCondition()</code> runs out of condition handlers before the condition is handled, it also returns normally.</p>
<p>The <code>stop()</code> function you’ve been using calls <code>signalCondition()</code>. If the error is handled by a condition handler that transfers control via <code>tryCatch()</code> or by invoking a restart, then the call to <code>signalCondition()</code> never returns.</p>
<p>Another condition signaling function, <code>warning()</code>, provides an example of a different kind of protocol built on the condition system. Like <code>stop()</code>, <code>warnings()</code> calls <code>signalCondition()</code> to signal a condition. But if <code>signalCondition()</code> returns, <code>warning()</code> doesn’t throw a top-level error–it prints the condition to <code>stderr</code> and returns <code>NULL</code>, allowing its caller to proceed. <code>warning()</code> also establishes a restart, <code>muffle_warning()</code>, around the call to <code>signalCondition()</code> that can be used by a condition handler to make <code>warning()</code> return without printing anything. Of course, a condition signaled with <code>warning()</code> could also be handled in some other way–a condition handler could “promote” a warning to an error by handling it as if it were an error.</p>
<p>For instance, in the log-parsing application, if there were ways a log entry could be slightly malformed but still parsable, you could write <code>parse_log_entry()</code> to go ahead and parse the slightly defective entries but to signal a condition with <code>warning()</code> when it did. Then the larger application could choose to let the warning print, to muffle the warning, or to treat the warning like an error, recovering the same way it would from a <code>malformed_log_entry_error</code>.</p>
<p>You can also build your own protocols on <code>signalCondition()</code>–whenever low-level code needs to communicate information back up the call stack to higher-level code, the condition mechanism is a reasonable mechanism to use. But for most purposes, one of the standard error or warning protocols should suffice.</p>
<p>Unfortunately, it’s the fate of error handling to always get short shrift in programming texts–proper error handling, or lack thereof, is often the biggest difference between illustrative code and hardened, production-quality code. The trick to writing the latter has more to do with adopting a particularly rigorous way of thinking about software than with the details of any particular programming language constructs. That said, if your goal is to write that kind of software, you’ll find the R condition system is an excellent tool for writing robust code and one that fits quite nicely into R’s incremental development style.</p>
<h3>Writing Robust Software</h3>
<p>For information on writing robust software, you could do worse than to start by finding a copy of <em>Software Reliability</em> (John Wiley &amp; Sons, 1976) by Glenford J. Meyers. Bertrand Meyer’s writings on Design By Contract also provide a useful way of thinking about software correctness. For instance, see Chapters 11 and 12 of his <em>Object-Oriented Software Construction</em> (Prentice Hall, 1997). Keep in mind, however, that Bertrand Meyer is the inventor of Eiffel, a statically typed bondage and discipline language in the Algol/Ada school. While he has a lot of smart things to say about object orientation and software reliability, there’s a fairly wide gap between his view of programming and The R Way. Finally, for an excellent overview of the larger issues surrounding building fault-tolerant systems, see Chapter 3 of the classic <em>Transaction Processing: Concepts and Techniques</em> (Morgan Kaufmann, 1993) by Jim Gray and Andreas Reuter.</p>
<p><sup>1</sup><em>Throws</em> or <em>raises</em> an exception in Java/Python terms</p>
<p><sup>2</sup><em>Catches</em> the exception in Java/Python terms</p>
<p><sup>3</sup>In this respect, a condition is a lot like an exception in Java or Python except not all conditions represent an error or <em>exceptional</em> situation.</p>
<p><sup>5</sup>The compiler may complain if the parameter is never used. You can silence that warning by adding a declaration <code>(declare (ignore c))</code> as the first expression in the <code>LAMBDA</code> body.</p>

        </div>
      </div>

      <div class="footer">
        <hr>
        <p>&copy; Hadley Wickham. Powered by <a href="http://jekyllrb.com/">jekyll</a>,
          <a href="http://yihui.name/knitr/">knitr</a>, and
          <a href="http://johnmacfarlane.net/pandoc/">pandoc</a>. Source
          available on <a href="https://github.com/hadley/adv-r/">github</a>.
        </p>
      </div>

    </div> <!-- /container -->

  <script src="//code.jquery.com/jquery.js"></script>
  <script src="www/bootstrap.min.js"></script>
  <script src="www/toc.js"></script>
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-67989-16', 'had.co.nz');
    ga('send', 'pageview');

  </script>
  </body>
</html>
